pkg/agent/plugin/workloadattestor/k8s/k8s_posix.go

```diff
@@ -565,22 +575,47 @@ func getPodUIDAndContainerIDFromCGroups(cgroups []cgroups.Cgroup) (types.UID, st
 	return podUID, containerID, nil
 }
 
-// cgroupRE is the regex used to parse out the pod UID and container ID from a
-// cgroup name. It assumes that any ".scope" suffix has been trimmed off
-// beforehand.  CAUTION: we used to verify that the pod and container id were
-// descendants of a kubepods directory, however, as of Kubernetes 1.21, cgroups
-// namespaces are in use and therefore we can no longer discern if that is the
-// case from within SPIRE agent container (since the container itself is
-// namespaced). As such, the regex has been relaxed to simply find the pod UID
-// followed by the container ID with allowances for arbitrary punctuation, and
-// container runtime prefixes, etc.
-var cgroupRE = regexp.MustCompile(`` +
-	// "pod"-prefixed Pod UID (with punctuation separated groups) followed by punctuation
-	`[[:punct:]]pod([[:xdigit:]]{8}[[:punct:]]?[[:xdigit:]]{4}[[:punct:]]?[[:xdigit:]]{4}[[:punct:]]?[[:xdigit:]]{4}[[:punct:]]?[[:xdigit:]]{12})[[:punct:]]` +
-	// zero or more punctuation separated "segments" (e.g. "docker-")
-	`(?:[[:^punct:]]+[[:punct:]])*` +
-	// non-punctuation end of string, i.e., the container ID
-	`([[:^punct:]]+)$`)
+var cgroupREs = []*regexp.Regexp{
+	// the regex used to parse out the pod UID and container ID from a
+	// cgroup name. It assumes that any ".scope" suffix has been trimmed off
+	// beforehand.  CAUTION: we used to verify that the pod and container id were
+	// descendants of a kubepods directory, however, as of Kubernetes 1.21, cgroups
+	// namespaces are in use and therefore we can no longer discern if that is the
+	// case from within SPIRE agent container (since the container itself is
+	// namespaced). As such, the regex has been relaxed to simply find the pod UID
+	// followed by the container ID with allowances for arbitrary punctuation, and
+	// container runtime prefixes, etc.
+	regexp.MustCompile(`` +
+		// "pod"-prefixed Pod UID (with punctuation separated groups) followed by punctuation
+		`[[:punct:]]pod(?P<poduid>[[:xdigit:]]{8}[[:punct:]]?[[:xdigit:]]{4}[[:punct:]]?[[:xdigit:]]{4}[[:punct:]]?[[:xdigit:]]{4}[[:punct:]]?[[:xdigit:]]{12})[[:punct:]]` +
+		// zero or more punctuation separated "segments" (e.g. "docker-")
+		`(?:[[:^punct:]]+[[:punct:]])*` +
+		// non-punctuation end of string, i.e., the container ID
+		`(?P<containerid>[[:^punct:]]+)$`),
+
+	// This regex applies for container runtimes, that won't put the PodUID into
+	// the cgroup name.
+	// Currently only cri-o in combination with kubeedge is known for this abnormaly.
+	regexp.MustCompile(`` +
+		// /crio-
+		`(?P<poduid>)[[:punct:]]crio[[:punct:]]` +
```

Comment:
good point.
Regex in gelang is quite limited. It would need a negative lookahead to not match, when a poduid is present.
I introduced a "mustnotmatch" group for this.